nexus/db-queries/src/db/datastore/instance.rs

```diff
@@ -1152,7 +1074,145 @@ impl DataStore {
                 public_error_from_diesel(e, ErrorHandler::Server)
             })?;
 
-        Ok(InstanceAndActiveVmm { instance, vmm })
+        Ok(instance_and_vmm)
+    }
+
+    pub async fn instance_set_boot_disk(
+        &self,
+        opctx: &OpContext,
+        authz_instance: &authz::Instance,
+        boot_disk_id: Option<Uuid>,
+    ) -> Result<(), Error> {
+        opctx.authorize(authz::Action::Modify, authz_instance).await?;
+
+        let err = OptionalError::new();
+        let conn = self.pool_connection_authorized(opctx).await?;
+        self.transaction_retry_wrapper("instance_set_boot_disk")
+            .transaction(&conn, |conn| {
+                let err = err.clone();
+                async move {
+                    self.instance_set_boot_disk_on_conn(
+                        &conn,
+                        &err,
+                        authz_instance,
+                        boot_disk_id,
+                    )
+                    .await?;
+                    Ok(())
+                }
+            })
+            .await
+            .map_err(|e| {
+                if let Some(err) = err.take() {
+                    return err;
+                }
+
+                public_error_from_diesel(e, ErrorHandler::Server)
+            })
+    }
+
+    /// Set an instance's boot disk to the provided `boot_disk_id` (or unset it,
+    /// if `boot_disk_id` is `None`), within an existing transaction.
+    ///
+    /// This is factored out as it is used by both
+    /// [`DataStore::instance_reconfigure`], which mutates many instance fields,
+    /// and [`DataStore::instance_set_boot_disk`], which only touches the boot
+    /// disk.
+    async fn instance_set_boot_disk_on_conn(
+        &self,
+        conn: &async_bb8_diesel::Connection<DbConnection>,
+        err: &OptionalError<Error>,
+        authz_instance: &authz::Instance,
+        boot_disk_id: Option<Uuid>,
+    ) -> Result<(), diesel::result::Error> {
+        use crate::db::model::InstanceState;
+
+        use db::schema::disk::dsl as disk_dsl;
+        use db::schema::instance::dsl as instance_dsl;
+
+        // * Allow reconfiguration in NoVmm because there is no VMM
+        //   to contend with.
+        // * Allow reconfiguration in Failed to allow changing the
+        //   boot disk of a failed instance and free its boot disk
+        //   for detach.
+        // * Allow reconfiguration in Creating because one of the
+        //   last steps of instance creation, while the instance is
+        //   still in Creating, is to reconfigure the instance to
+        //   the desired boot disk.
+        let ok_to_reconfigure_instance_states = [
+            InstanceState::NoVmm,
+            InstanceState::Failed,
+            InstanceState::Creating,
+        ];
```

Comment:
@iximeow I'm considering pulling this check out of `instance_set_boot_disk_on_conn` and putting it in the top of `instance_reconfigure`'s transaction. That way, we can remove the `Creating` state from the list of allowed states, since the `instance_create` saga no longer uses `instance_reconfigure` but instead calls into the `instance_set_boot_disk` method directly. That way, we don't have to deal with the potentially sketchy consequences of allowing API-driven reconfiguration of instances while they are being created. WDYT?

Reply:
yeah, that seems pretty reasonable to me!

Reply:
Done in 1c9800fde93efc925c13ce4009be260e3b2e4af0.

There is a caveat that we could, potentially, allow updates in other states if the boot-disk part of the update is a no-op, but doing that correctly becomes kind of annoying. I left a note about it --- we _could_ do it here but it feels high effort.

On the other hand, it does seem desirable to be able to set auto-restart policies for running instances...